lib/fluent/plugin/buffer.rb

```diff
@@ -0,0 +1,351 @@
+          @total_bytes_limit = @chunk_bytes_limit * @queue_length_limit
+        end
+      end
+
+      def start
+        super
+        # stage #=> Hash (metadata -> chunk): not flushed yet
+        # queue #=> Array (chunks)          : already flushed (not written)
+        @stage, @queue = resume
+
+        @queued_num = {} # metadata => int (number of queued chunks)
+        @queue.each do |chunk|
+          @queued_num[chunk.metadata] ||= 0
+          @queued_num[chunk.metadata] += 1
+        end
+
+        @dequeued = {} # unique_id => chunk
+
+        @stage_size = @queue_size = 0
+        @metadata_list = [] # keys of @stage
+      end
+
+      def close
+        super
+        synchronize do
+          @dequeued.each_pair do |chunk_id, chunk|
+            chunk.close
+          end
+          until @queue.empty?
+            @queue.shift.close
+          end
+          @stage.each_pair do |metadata, chunk|
+            chunk.close
+          end
+        end
+      end
+
+      def terminate
+        super
+        @stage = @queue = nil
+      end
+
+      def storable?
+        @total_size_limit > @stage_size + @queue_size
+      end
+
+      ## TODO: for back pressure feature
+      # def used?(ratio)
+      #   @total_size_limit * ratio > @stage_size + @queue_size
+      # end
+
+      def resume
+        # return {}, []
+        raise NotImplementedError, "Implement this method in child class"
+      end
+
+      def generate_chunk(metadata)
+        raise NotImplementedError, "Implement this method in child class"
+      end
+
+      def metadata_list
+        synchronize do
+          @metadata_list.dup
+        end
+      end
+
+      def new_metadata(timekey: nil, tag: nil, variables: nil)
+        Metadata.new(timekey, tag, variables)
+      end
+
+      # TODO: metadata_list cleaner
+      def add_metadata(metadata)
+        synchronize do
+          if i = @metadata_list.index(metadata)
+            @metadata_list[i]
+          else
+            @metadata_list << metadata
+          end
+        end
+      end
+
+      def metadata(timekey: nil, tag: nil, variables: nil)
+        meta = new_meatadata(timekey: timekey, tag: tag, variables: variables)
+        add_metadata(meta)
+        meta
+      end
+
+      # metadata MUST have consistent object_id for each variation
+      # data MUST be Array of serialized events
+      def emit(metadata, data, force: false)
+        data_size = data.size
+        return if data_size < 1
+        raise BufferOverflowError unless storable?
+
+        stored = false
+
+        # the case whole data can be stored in staged chunk: almost all emits will success
+        chunk = synchronize { @stage[metadata] ||= generate_chunk(metadata) }
+        chunk.synchronize do
+          begin
+            chunk.append(data)
+            if !size_over?(chunk) || force
+              chunk.commit
+              stored = true
+            end
+          ensure
```

Comment:
NO. `ensure`. This code always do `rollback` to reset internal state perfectly.
With `rescue`, this code should be like this:
```ruby
        chunk.synchronize do
          begin
            chunk.append(data)
            if !size_over?(chunk) || force
              chunk.commit
              stored = true
            else
              chunk.rollback
            end
          rescue
            chunk.rollback
            raise
          end
        end
```
This code is a bit efficient, but more complex.